### ISSUE TITLE
Problem building installer (following the tutorial)

### ISSUE THREAD

**Issue by hellbe** [mherrmann/fbs#33]
Followed tutorial to the point. 

- Windows 10
- Python 3.6 x64
- NSIS 3.03

**Output:**
```
(venv) C:\Users\myuser\code\testapp>python -m fbs installer
Processing config: C:\Program Files (x86)\NSIS\nsisconf.nsh
Processing script file: "Installer.nsi" (ACP)
File: "..\TestApp\*" -> no files found.
Usage: File [/nonfatal] [/a] ([/r] [/x filespec [...]] filespec [...] |
   /oname=outfile one_file_only)
Error in script "Installer.nsi" on line 70 -- aborting creation process
Traceback (most recent call last):
  File "c:\users\myuser\appdata\local\programs\python\python36\Lib\runpy.py", line 193, in _run_module_as_main
    "__main__", mod_spec)
  File "c:\users\myuser\appdata\local\programs\python\python36\Lib\runpy.py", line 85, in _run_code
    exec(code, run_globals)
  File "C:\Users\myuser\code\testapp\venv\lib\site-packages\fbs\__main__.py", line 4, in <module>
    main()
  File "C:\Users\myuser\code\testapp\venv\lib\site-packages\fbs\cmdline.py", line 25, in main
    args.fn(*fn_args)
  File "C:\Users\myuser\code\testapp\venv\lib\site-packages\fbs\builtin_commands\__init__.py", line 94, in installer
    create_installer_windows()
  File "C:\Users\myuser\code\testapp\venv\lib\site-packages\fbs\installer\windows.py", line 19, in create_installer_windows
    run(['makensis', 'Installer.nsi'], cwd=path('target/NSIS'), check=True)
  File "c:\users\myuser\appdata\local\programs\python\python36\Lib\subprocess.py", line 418, in run
    output=stdout, stderr=stderr)
subprocess.CalledProcessError: Command '['makensis', 'Installer.nsi']' returned non-zero exit status 1.
```

**Installer.nsi:**
```
!include MUI2.nsh
!include FileFunc.nsh

;--------------------------------
;Perform Machine-level install, if possible

!define MULTIUSER_EXECUTIONLEVEL Highest
;Add support for command-line args that let uninstaller know whether to
;uninstall machine- or user installation:
!define MULTIUSER_INSTALLMODE_COMMANDLINE
!include MultiUser.nsh
!include LogicLib.nsh

Function .onInit
  !insertmacro MULTIUSER_INIT
  ;Do not use InstallDir at all so we can detect empty $InstDir!
  ${If} $InstDir == "" ; /D not used
      ${If} $MultiUser.InstallMode == "AllUsers"
          StrCpy $InstDir "$PROGRAMFILES\TestApp"
      ${Else}
          StrCpy $InstDir "$LOCALAPPDATA\TestApp"
      ${EndIf}
  ${EndIf}
FunctionEnd

Function un.onInit
  !insertmacro MULTIUSER_UNINIT
FunctionEnd

;--------------------------------
;General

  Name "TestApp"
  OutFile "..\TestAppSetup.exe"

;--------------------------------
;Interface Settings

  !define MUI_ABORTWARNING

;--------------------------------
;Pages

  !define MUI_WELCOMEPAGE_TEXT "This wizard will guide you through the installation of TestApp.$\r$\n$\r$\n$\r$\nClick Next to continue."
  !insertmacro MUI_PAGE_WELCOME
  !insertmacro MUI_PAGE_DIRECTORY
  !insertmacro MUI_PAGE_INSTFILES
    !define MUI_FINISHPAGE_NOAUTOCLOSE
    !define MUI_FINISHPAGE_RUN
    !define MUI_FINISHPAGE_RUN_CHECKED
    !define MUI_FINISHPAGE_RUN_TEXT "Run TestApp"
    !define MUI_FINISHPAGE_RUN_FUNCTION "LaunchLink"
  !insertmacro MUI_PAGE_FINISH

  !insertmacro MUI_UNPAGE_CONFIRM
  !insertmacro MUI_UNPAGE_INSTFILES

;--------------------------------
;Languages

  !insertmacro MUI_LANGUAGE "English"

;--------------------------------
;Installer Sections

!define UNINST_KEY \
  "Software\Microsoft\Windows\CurrentVersion\Uninstall\TestApp"
Section
  SetOutPath "$InstDir"
  File /r "..\TestApp\*"
  WriteRegStr SHCTX "Software\TestApp" "" $InstDir
  WriteUninstaller "$InstDir\uninstall.exe"
  CreateShortCut "$SMPROGRAMS\TestApp.lnk" "$InstDir\TestApp.exe"
  WriteRegStr SHCTX "${UNINST_KEY}" "DisplayName" "TestApp"
  WriteRegStr SHCTX "${UNINST_KEY}" "UninstallString" \
    "$\"$InstDir\uninstall.exe$\" /$MultiUser.InstallMode"
  WriteRegStr SHCTX "${UNINST_KEY}" "QuietUninstallString" \
    "$\"$InstDir\uninstall.exe$\" /$MultiUser.InstallMode /S"
  WriteRegStr SHCTX "${UNINST_KEY}" "Publisher" "Simon Hellbe"
  ${GetSize} "$InstDir" "/S=0K" $0 $1 $2
  IntFmt $0 "0x%08X" $0
  WriteRegDWORD SHCTX "${UNINST_KEY}" "EstimatedSize" "$0"

SectionEnd

;--------------------------------
;Uninstaller Section

Section "Uninstall"

  RMDir /r "$InstDir"
  Delete "$SMPROGRAMS\TestApp.lnk"
  DeleteRegKey /ifempty SHCTX "Software\TestApp"
  DeleteRegKey SHCTX "${UNINST_KEY}"

SectionEnd

Function LaunchLink
  !addplugindir "."
  ShellExecAsUser::ShellExecAsUser "open" "$SMPROGRAMS\TestApp.lnk"
FunctionEnd
```

**Comment by mherrmann** [comment]
To the point - except using Python 3.**5** ;)

**Comment by hellbe** [comment]
*facepalm* .. will try with 3.5

Based on previous experience I suspect 3.5 is a PyInstaller limitation?

**Comment by mherrmann** [comment]
Exactly, yes. If you're interested in blood, sweat and tears see [here](https://fman.io/blog/battling-with-macos/).

**Comment by mherrmann** [comment]
Oh and did you run `freeze` before `installer`? That's required as well.

**Comment by hellbe** [comment]
@mherrmann yep, after removing part of Jinja2 I am getting freeze to run properly now. Unfortunately the dist fails with unknown error. How can I troubleshoot?

Would like to create a user files install that doesn't require admin permissions. Any pointers?

**Comment by hellbe** [comment]
New error:
```
....
....
[17992] LOADER: Running main.py
Traceback (most recent call last):
  File "site-packages\fbs_runtime\_signal.py", line 2, in <module>
  File "c:\users\sihel7\code\testapp\venv\lib\site-packages\PyInstaller\loader\pyimod03_importers.py", line 714, in load_module
    module = loader.load_module(fullname)
ImportError: No module named 'PyQt5.sip'

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "main.py", line 1, in <module>
  File "c:\users\sihel7\code\testapp\venv\lib\site-packages\PyInstaller\loader\pyimod03_importers.py", line 631, in exec_module
    exec(bytecode, module.__dict__)
  File "site-packages\fbs_runtime\application_context.py", line 2, in <module>
  File "c:\users\sihel7\code\testapp\venv\lib\site-packages\PyInstaller\loader\pyimod03_importers.py", line 631, in exec_module
    exec(bytecode, module.__dict__)
  File "site-packages\fbs_runtime\_signal.py", line 4, in <module>
ImportError: No module named 'PySide2'
[17992] Failed to execute script main
[17992] LOADER: OK.
[17992] LOADER: Cleaning up Python interpreter.
```

**Comment by hellbe** [comment]
Maybe this: https://stackoverflow.com/a/51330788

**Comment by mherrmann** [comment]
I'm surprised - that's happening with Python 3.5?

Regarding user install that doesn't require admin permissions: The way it's [set up](https://github.com/mherrmann/fbs/blob/5af1c18448216e77fa8a7f95feadd459b2a9b10d/fbs/builtin_commands/project_template/src/main/NSIS/Installer.nsi#L7), it should attempt to perform an admin install, if that fails because of insufficient permissions, it performs a user install. If you want to force it to always perform a user install, you should be able to set `MULTIUSER_EXECUTIONLEVEL` to `Standard`.

**Comment by mherrmann** [comment]
`python -m fbs run` works?

**Comment by hellbe** [comment]
Yes!

**Comment by mherrmann** [comment]
The exact same error was recently encountered by @boulund (see [here](https://github.com/mherrmann/fbs/issues/31#issuecomment-425711139)). Time I look into this... Do you maybe have a very minimal app using pandas?

**Comment by boulund** [comment]
I built a small example that produces the issue:

https://github.com/boulund/fbs_pandas_example

**Comment by mherrmann** [comment]
Thank you. I'll take a loo.

**Comment by mherrmann** [comment]
Okay, here's how to fix this. Add the following lines to [`main.py`](https://github.com/boulund/fbs_pandas_example/blob/master/src/main/python/main.py):

    import pandas._libs.tslibs.np_datetime
    import pandas._libs.tslibs.nattype
    import pandas._libs.skiplist

How did I come up with this? I looked at the error you posted above:

    ImportError: C extension: No module named 'pandas._libs.tslibs.np_datetime' not built. [...]

So I added `import pandas._libs.tslibs.np_datetime`. Then I got a similar error for `...nattype` so I added the second import. And so on.

In summary:
 * When you get **Failed to execute script main**, use `freeze --debug` to get more info.
 * When faced with an `ImportError` when running the frozen form of your app, add the corresponding `import` statement to your code.

An alternative is to use PyInstaller's `--hidden-import` flag. But that requires a [custom build script](https://build-system.fman.io/manual/#custom-commands). If you want me to explain this, let me know.

**Comment by mherrmann** [comment]
I may add a `hidden_imports` setting to [`base.json`](https://github.com/mherrmann/fbs/blob/master/fbs/default_settings/base.json) to let you do this more elegantly. What do you think?

**Comment by boulund** [comment]
Nice! Thank so much for spending the time! From what I read when googling around it seemed to be related to some hidden imports, but I never got the time to really sit down and have a proper look. You saved me a day of work rewriting all table import code in my apps 👍 

Having a `hidden_imports` setting in `base.json` would be probably be a clean way to expose this, yes. But what would you think is the best way to find the hidden imports? Just freezing and trying to run the application?

**Comment by mherrmann** [comment]
My pleasure :-)

> what would you think is the best way to find the hidden imports? Just freezing and trying to run the application, iteratively adding hidden imports?

Yes, that's how I would do it. It sounds painful. But given how many libraries there are out there, fbs can't know the hidden imports for all of them. Plus, you only have to do this once. I'll expand the [documentation](https://build-system.fman.io/manual/#dependencies) to explain this.

I'll look into a `hidden_imports` setting. The only thing I'm not sure about is how I should handle `hidden_imports` defined in multiple settings. For instance: It may be defined in `base.json` and `mac.json`.  Should the latter _extend_ or _overwrite_ the former? :thinking:

**Comment by boulund** [comment]
Tricky question. I'm not (yet) intimately familiar with the structure and layout of the fman build system, so I really can't tell, sorry! Maybe it's just easiest to explicitly import them like I did now?

**Comment by mherrmann** [comment]
No problem. I'm not sure either yet. It's a common use case to have to add hidden imports. (fman does it as well.) So would be nice to have a solution. I'll expand the docs first, then keep thinking about it.

**Comment by mherrmann** [comment]
[docs](https://build-system.fman.io/manual/#dependencies) updated.

**Comment by hellbe** [comment]
Awesome, just wanted to say thanks @mherrmann and @boulund!

**Comment by mherrmann** [comment]
Happy to help :-)
